Unified Storage: Don't read before create

Proximyst opened the pull request:
Follow up to #102857. We don't need to read before creating as the database will tell us if something already exists.

Proximyst committed:
Unified Storage: Don't read before create

marcoabreu commented on pkg/storage/unified/resource/server.go:
```diff
@@ -476,22 +476,14 @@ func (s *server) Create(ctx context.Context, req *CreateRequest) (*CreateRespons
 	}
 
+	// If the resource already exists, the create will return an already exists error that is remapped appropriately by AsErrorResult.
+	// This also benefits from ACID behaviours on our databases, so we avoid race conditions.
```
Super important, great improvement!

marcoabreu commented on pkg/storage/unified/store_test.go:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	t.Run("happy path: creates a new resource", func(t *testing.T) {
+		// TODO: Make a Spanner-compatible test
```
If you just forget about everything related to SQL and just work through the API to create objects, then that should be totally sufficient.

In an integration test for a storage backend, you'd basically want to watch out for the word "SQL" appearing anywhere in your code and then reconsider the approach. Because the backend could be anything, so the test must not be aware of SQL.

We already do have integration tests, couldn't you just adopt their approach?

Proximyst replied:
Absolutely can :) I aimed to just validate the code would work at all in an easy manner. I'll migrate to a proper integration test within that setup.

Proximyst committed:
Merge remote-tracking branch 'origin/main' into mariell/remove-read-before-create

Proximyst committed:
test: use the existing test infra

Proximyst committed:
fix: support pq

We use pgx, but it seems to be wrapped in a pq driver shim, causing the errors to be remapped to pq's type. Weird
situation.

Proximyst committed:
feat: support CDK backend

JohnnyQQQQ commented on devenv/docker/blocks/postgres/docker-compose.yaml:
```diff
@@ -1,9 +1,9 @@
-      POSTGRES_USER: grafana
-      POSTGRES_PASSWORD: password
-      POSTGRES_DB: grafana
+      POSTGRES_USER: grafanatest
+      POSTGRES_PASSWORD: grafanatest
+      POSTGRES_DB: grafanatest
```
Can we please revert this change and keep it consistent with the other blocks.

Proximyst replied:
ack, I see now I missed the `postgres_tests` block... will revert this :)

Proximyst committed:
revert: there is a postgres_tests block

Proximyst committed:
fix(CDK): only check existence on ADDED updates

Proximyst committed:
fix(CDK): use ReadResource to deal with deleted files

ryantxu reviewed (approved):
nice
